ipywidgets/widgets/trait_types/eventful.py

```diff
@@ -0,0 +1,180 @@
+from .spectate import watch, unwatch, watcher, expose_as, watchable
+from traitlets import TraitType, Undefined
+from traitlets.utils.bunch import Bunch
+
+
+class Callback(object):
+
+    def __init__(self, owner, trait, etype, callback):
+        self.owner = owner
+        self.trait = trait
+        self.etype = etype
+        self.func = callback
+
+    def __eq__(self, other):
+        if not isinstance(other, Callback):
+            return self.func == other
+        else:
+            return other is self
+
+
+class Beforeback(Callback):
+
+    def __call__(self, inst, call):
+        call = call.copy()
+        call.update(
+            trait=self.trait,
+            owner=self.owner,
+            type=self.etype)
+        return self.func(inst, call)
+
+
+class Afterback(Callback):
+
+    def __call__(self, inst, answer):
+        if self.func is None:
+            result = answer['before']
+            if (isinstance(result, Redirect)
+                and result.origin is not None):
+                origin = result.origin
+            else:
+                origin = self.etype
+
+            event = (result(answer.value) if
+                callable(result) else result)
+        else:
+            origin = self.etype
+            event = self.func(inst, answer)
+        if None not in (origin, event):
+            self.owner.notify_change(Bunch(
+                type=origin, event=event,
+                name=self.trait.name,
+                owner=self.owner))
+        return event
+
+
+class Redirect(object):
+
+    def __init__(self, origin, target, inst, args, kwargs):
+        self.origin = origin
+        registry = inst._instance_spectator._callback_registry
+        # Eventful._setup_events registers first
+        b, a = registry.get(target, [(None, None)])[0]
```

Comment:
b and a force me to go read the docs for `registry.get`. Can you use names here?